### PR TITLE
ARROW-5533: [C++] [Plasma] make plasma client thread safe

### PR DESCRIPTION
**Pull request by zhijunfu**
Make plasma client thread safe, so it's guaranteed when the PlasmaBuffer for an object destructs, the call to plasma_client.Release(object_id) is properly protected.  Refer to [here](https://github.com/ray-project/ray/pull/4922#discussion_r291422782) for background.

@robertnishihara @pcmoritz

### CODE REVIEW

**Comment by robertnishihara** [comment]
This looks good to me, but it seems like there might be some related test failures in Travis.

**Comment by zhijunfu** [comment]
@robertnishihara all travis tests pass now. I changed `mutex` to `recursive_mutex` as some plasma client interfaces are called from other ones, thus the lock needs to be reentrant. 
I'll also file a PR on the ray side to verify the change.

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
LGTM, but let's wait for the Ray test suite to pass before merging it!

**Comment by pcmoritz** [comment]
rebased to include https://github.com/apache/arrow/pull/4494

File: cpp/src/plasma/client.cc [hunk 1]
```diff
@@ -503,7 +510,7 @@ Status PlasmaClient::Impl::GetBuffers(
             data + object->data_offset, object->data_size + object->metadata_size);
       } else {
 #ifdef PLASMA_CUDA
-        std::lock_guard<std::mutex> lock(gpu_mutex);
+        std::lock_guard<std::recursive_mutex> lock(gpu_mutex);
```
Comment (shengjun1985) [inline cpp/src/plasma/client.cc]: It will cause compile error because of the definition "static std::mutex gpu_mutex;" at line 110.

**Review by shengjun1985 — CHANGES_REQUESTED** [review]

File: cpp/src/plasma/client.cc (hunk 1 above)
Comment (zhijunfu) [reply cpp/src/plasma/client.cc]: @shengjun1985 Thanks for catching this! It's fixed.

**Review by shengjun1985 — COMMENTED** [review]
Thank you for fix, but I have no idea why CI failed.

**Comment by zhijunfu** [comment]
@shengjun1985 that one is specific to mac and seems unrelated.

**Review by shengjun1985 — APPROVED** [review]

**Comment by wesm** [comment]
Merging. CI failure looks unrelated to me

**Pull request closed without merging**